### DISCUSSION TITLE
pre-commit unsupported version values (black/mypy/pylint)

### DISCUSSION THREAD

**Discussion by monrad** [renovatebot/renovate#9439] [Request Help] [answered]
I just started testing Renovate, awesome project by the way :), and when running the pre-commit manager I get some errors that some deps have unsupported values:
```
  "msg": "Dependency ambv/black has unsupported value 20.8b1",
  "msg": "Dependency pycqa/pylint has unsupported value pylint-2.7.4",
  "msg": "Dependency pre-commit/mirrors-mypy has unsupported value v0.812",
```
Should I match them with packagePatterns and use versionScheme to match them or is there a better way?

**Comment by rarkins** [comment, marked as answer]
Yes, you will need to define package rules to match them and override `versioning`. For `ambv/black` it looks like you can set versioning to `pep440` and maybe also for `pre-commit/mirrors-mypy` too. Do you know why `pyqa/pylint` has the `pylint-` prefix in its version? You may need to use `extractVersion` in your config to work around that.

**Reply by monrad** [reply]
Thanks,
For black and mypy it was easy enough, this config works like a charm:
```
"packageRules": [
    {
      "matchPackageNames": [
        "ambv/black",
        "pre-commit/mirrors-mypy"
      ],
      "matchManagers": [
        "pre-commit"
      ],
      "versioning": "pep440"
    }
]
```
But I have a bit of an issue with extractVersion
```
  "packageRules": [
    {
      "matchPackageNames": [
        "pycqa/pylint"
      ],
      "matchManagers": [
        "pre-commit"
      ],
      "extractVersion": "^pylint-(?<version>.*)$"
    }
  ]
```
This does not seem to work, do I need to match the datasource or something else?
I did try pylint with and without the matchMangers.
I have no idea why they do it but in all their GitHub release version names they have included "pylint-" for some reason.

**Reply by rarkins** [reply]
It looks like it should be matching fine, in which case the `extractVersion` is maybe not working. Can you create a reproduction repo with just that one pre-commit dependency to debug against?

**Reply by monrad** [reply]
I have created https://github.com/monrad/renovate-extractversion-pylint and it also shows "Dependency pycqa/pylint has unsupported value pylint-2.7.4" should I create a bug report instead of this discussion?

**Reply by rarkins** [reply]
No need, it may be solvable by configuration. It's not a bug at all that Renovate by default doesn't update it - only a bug if extractVersion is not working as intended

**Reply by rarkins** [reply]
Actually, it's getting rejected prior to `extractVersion` being used or the datasource being called. This might require regex versioning instead

**Reply by monrad** [reply]
That seems to work, no more unsupported in the log.
I just tested to see if it would upgrade with a test [repo](https://github.com/monrad/renovate-regex-versioning-pylint/pull/1/files) and it works, only minor thing is it writes "Update precommit hook pycqa/pylint to vpylint-2.7.4" putting the v in front of pylint in the PR name

This is my full current config if anybody is googling something similar :)
```
{
  "packageRules": [
    {
      "matchPackageNames": [
        "pycqa/pylint"
      ],
      "matchManagers": [
        "pre-commit"
      ],
      "versioning": "regex:^pylint-(?<major>\\d+)\\.(?<minor>\\d+)\\.(?<patch>\\d+)$"
    },
    {
      "matchPackageNames": [
        "ambv/black",
        "pre-commit/mirrors-mypy"
      ],
      "matchManagers": [
        "pre-commit"
      ],
      "versioning": "pep440"
    }
  ]
}
```

@rarkins Thanks for your help and making such a great tool, I hope you have a great day.
